src/keyboard/mod.rs

```diff
@@ -569,59 +612,252 @@ where
                     let sym = state.get_one_sym_raw(key);
-                    let ignore_text = if key_state == wl_keyboard::KeyState::Pressed {
-                        state.compose_feed(sym)
+                    let utf8 = { 
+                        if state.compose_feed(sym)
                             != Some(ffi::xkb_compose_feed_result::XKB_COMPOSE_FEED_ACCEPTED)
-                    } else {
-                        true
-                    };
-                    let utf8 = if ignore_text {
-                        None
-                    } else if let Some(status) = state.compose_status() {
-                        match status {
-                            ffi::xkb_compose_status::XKB_COMPOSE_COMPOSED => {
-                                state.compose_get_utf8()
+                        {
+                            None
+                        } else if let Some(status) = state.compose_status() {
+                            match status {
+                                ffi::xkb_compose_status::XKB_COMPOSE_COMPOSED => {
+                                    state.compose_get_utf8()
+                                }
+                                ffi::xkb_compose_status::XKB_COMPOSE_NOTHING => state.get_utf8_raw(key),
+                                _ => None,
                             }
-                            ffi::xkb_compose_status::XKB_COMPOSE_NOTHING => state.get_utf8_raw(key),
-                            _ => None,
+                        } else {
+                            state.get_utf8_raw(key)
                         }
-                    } else {
-                        state.get_utf8_raw(key)
                     };
                     let modifiers = state.mods_state.clone();
-                    user_impl.receive(
-                        Event::Key {
-                            serial,
-                            time,
-                            modifiers,
-                            rawkey: key,
-                            keysym: sym,
-                            state: key_state,
-                            utf8,
-                        },
-                        proxy,
-                    );
+
+                    if key_state == wl_keyboard::KeyState::Pressed {
+                        event_impl.receive(
+                            Event::Key {
+                                serial,
+                                time,
+                                modifiers,
+                                rawkey: key,
+                                keysym: sym,
+                                state: key_state,
+                                utf8: utf8.clone(),
+                            },
+                            proxy.clone(),
+                        );
+                        if let Some(repeat_impl) = repeat_impl.clone() {
+                            // Check with xkb if key is repeatable
+                            if unsafe { state.key_repeats(key + 8) } {
```

Comment:
the `+ 8` offsetting should happen inside of the `key_repeats` method, like for the other methods of `KbState`.